File: results/dnd-rest-benchmark/lifecycle-runs/20260710T233736Z_pi_kimi-k2p7-code_typescript-nextjs/app/lib/combat.ts
import { getDb } from "./db.js";

export interface CombatantInput {
  name: string;
  dex: number;
  roll: number;
}

export interface CombatantOrder {
  name: string;
  score: number;
}

export interface Condition {
  condition: string;
  remaining_rounds: number;
}

export interface CombatSession {
  id: string;
  round: number;
  turn_index: number;
  order: CombatantOrder[];
  conditions: Record<string, Condition[]>;
}

function rowToSession(
  row: Record<string, unknown> | undefined
): CombatSession | undefined {
  if (!row) return undefined;
  return {
    id: String(row.id),
    round: Number(row.round),
    turn_index: Number(row.turn_index),
    order: JSON.parse(String(row.order_json)),
    conditions: JSON.parse(String(row.conditions_json)),
  };
}

function getSessionById(id: string): CombatSession | undefined {
  const db = getDb();
  const row = db
    .prepare(
      "SELECT id, round, turn_index, order_json, conditions_json FROM combat_sessions WHERE id = ?"
    )
    .get(id);
  return rowToSession(row);
}

function saveSession(session: CombatSession): void {
  const db = getDb();
  db.prepare(
    `
    INSERT INTO combat_sessions (id, round, turn_index, order_json, conditions_json)
    VALUES (?, ?, ?, ?, ?)
    ON CONFLICT(id) DO UPDATE SET
      round = excluded.round,
      turn_index = excluded.turn_index,
      order_json = excluded.order_json,
      conditions_json = excluded.conditions_json
    `
  ).run(
    session.id,
    session.round,
    session.turn_index,
    JSON.stringify(session.order),
    JSON.stringify(session.conditions)
  );
}

export function createSession(
  id: unknown,
  combatants: unknown
): CombatSession {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("id must be a non-empty string");
  }
  if (!Array.isArray(combatants) || combatants.length === 0) {
    throw new Error("combatants must be a non-empty array");
  }

  const seen = new Set<string>();
  for (const c of combatants) {
    if (typeof c !== "object" || c === null) {
      throw new Error("each combatant must be an object");
    }
    const combatant = c as Record<string, unknown>;
    if (typeof combatant.name !== "string" || combatant.name.length === 0) {
      throw new Error("combatant name must be a non-empty string");
    }
    if (seen.has(combatant.name)) {
      throw new Error("combatant names must be unique");
    }
    seen.add(combatant.name);
    if (
      typeof combatant.dex !== "number" ||
      !Number.isInteger(combatant.dex)
    ) {
      throw new Error("combatant dex must be an integer");
    }
    if (
      typeof combatant.roll !== "number" ||
      !Number.isInteger(combatant.roll)
    ) {
      throw new Error("combatant roll must be an integer");
    }
  }

  const db = getDb();
  const existing = db
    .prepare("SELECT id FROM combat_sessions WHERE id = ?")
    .get(id);
  if (existing) {
    throw new Error("session id already exists");
  }

  const order = (combatants as CombatantInput[])
    .map((c) => ({ name: c.name, score: c.roll + c.dex, dex: c.dex }))
    .sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      if (b.dex !== a.dex) return b.dex - a.dex;
      return a.name.localeCompare(b.name);
    })
    .map(({ name, score }) => ({ name, score }));

  const session: CombatSession = {
    id,
    round: 1,
    turn_index: 0,
    order,
    conditions: {},
  };

  saveSession(session);
  return session;
}

export function addCondition(
  sessionId: string,
  target: unknown,
  condition: unknown,
  durationRounds: unknown
): { target: string; conditions: Condition[] } {
  const session = getSessionById(sessionId);
  if (!session) {
    throw new Error("session not found");
  }

  if (typeof target !== "string" || target.length === 0) {
    throw new Error("target must be a non-empty string");
  }
  if (!session.order.some((c) => c.name === target)) {
    throw new Error("target must be a combatant in the session");
  }
  if (typeof condition !== "string") {
    throw new Error("condition must be a string");
  }
  if (
    typeof durationRounds !== "number" ||
    !Number.isInteger(durationRounds) ||
    durationRounds <= 0
  ) {
    throw new Error("duration_rounds must be a positive integer");
  }

  const newCondition: Condition = {
    condition,
    remaining_rounds: durationRounds,
  };
  session.conditions[target] = [
    ...(session.conditions[target] ?? []),
    newCondition,
  ];

  saveSession(session);
  return { target, conditions: session.conditions[target] };
}

export function advanceTurn(sessionId: string): {
  id: string;
  round: number;
  turn_index: number;
  active: CombatantOrder;
  conditions: Record<string, Condition[]>;
} {
  const session = getSessionById(sessionId);
  if (!session) {
    throw new Error("session not found");
  }

  session.turn_index += 1;
  if (session.turn_index >= session.order.length) {
    session.turn_index = 0;
    session.round += 1;
  }

  const activeName = session.order[session.turn_index].name;
  const activeConditions = session.conditions[activeName] ?? [];
  const remaining = activeConditions
    .map((c) => ({ ...c, remaining_rounds: c.remaining_rounds - 1 }))
    .filter((c) => c.remaining_rounds > 0);

  if (activeConditions.length > 0) {
    session.conditions[activeName] = remaining;
  } else {
    delete session.conditions[activeName];
  }

  saveSession(session);

  return {
    id: session.id,
    round: session.round,
    turn_index: session.turn_index,
    active: session.order[session.turn_index],
    conditions: session.conditions,
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_rust-stdlib/src/combat.rs
use std::collections::HashMap;

use crate::domain::{CombatSession, Combatant, Condition};
use crate::json::{escape_json, extract_array_content, extract_int, extract_objects, extract_string};

pub enum CombatError {
    NotFound,
    BadRequest,
}

/// Parse a session id from a path like `/v1/combat/sessions/{id}/suffix`.
pub fn parse_combat_path<'a>(path: &'a str, suffix: &'a str) -> Option<&'a str> {
    let rest = path.strip_prefix("/v1/combat/sessions/")?;
    let id = rest.strip_suffix(suffix)?;
    if id.is_empty() {
        return None;
    }
    Some(id)
}

/// Build a list of combatants from a JSON array body and sort them into
/// initiative order.
///
/// The score is `roll + dex`. Ties are broken by higher dex, then by name in
/// ascending lexicographic order to keep the result deterministic.
fn roll_initiative(combatants_body: &str) -> Option<Vec<Combatant>> {
    let objects = extract_objects(combatants_body);
    if objects.is_empty() {
        return None;
    }
    let mut order = Vec::new();
    for obj in objects {
        let name = extract_string(obj, "name")?.to_string();
        let dex = extract_int(obj, "dex")?;
        let roll = extract_int(obj, "roll")?;
        order.push(Combatant {
            name,
            score: roll + dex,
            dex,
        });
    }
    order.sort_by(|a, b| {
        b.score
            .cmp(&a.score)
            .then(b.dex.cmp(&a.dex))
            .then(a.name.cmp(&b.name))
    });
    Some(order)
}

/// Create a new combat session from a request body.
///
/// Required fields: `id` (string), `combatants` (array of objects with
/// `name`, `dex`, `roll`). The session starts at round 1, turn 0, with the
/// top initiative combatant active. Returns the full session summary.
pub fn create_session(
    body: &str,
    sessions: &mut HashMap<String, CombatSession>,
) -> Option<String> {
    let id = extract_string(body, "id")?.to_string();
    if id.is_empty() || sessions.contains_key(&id) {
        return None;
    }
    let content = extract_array_content(body, "combatants")?;
    let order = roll_initiative(content)?;
    let active = order.first()?.clone();
    sessions.insert(
        id.clone(),
        CombatSession {
            id: id.clone(),
            round: 1,
            turn_index: 0,
            order: order.clone(),
            conditions: HashMap::new(),
        },
    );
    Some(format_session_summary(&id, 1, 0, &active, &order))
}

/// Serialize a session summary in the format expected by the API.
fn format_session_summary(
    id: &str,
    round: i64,
    turn_index: usize,
    active: &Combatant,
    order: &[Combatant],
) -> String {
    let order_json = order
        .iter()
        .map(|c| format!(r#"{{"name":"{}","score":{}}}"#, escape_json(&c.name), c.score))
        .collect::<Vec<_>>()
        .join(",");
    format!(
        r#"{{"id":"{}","round":{},"turn_index":{},"active":{{"name":"{}","score":{}}},"order":[{}]}}"#,
        escape_json(id),
        round,
        turn_index,
        escape_json(&active.name),
        active.score,
        order_json
    )
}

/// Add a condition to a target combatant in a session.
///
/// The target must exist in the session order. `duration_rounds` must be a
/// positive integer. Returns the updated condition list for that target.
pub fn add_condition(
    id: &str,
    body: &str,
    sessions: &mut HashMap<String, CombatSession>,
) -> Result<String, CombatError> {
    let session = sessions.get_mut(id).ok_or(CombatError::NotFound)?;
    let target = extract_string(body, "target")
        .ok_or(CombatError::BadRequest)?
        .to_string();
    if !session.order.iter().any(|c| c.name == target) {
        return Err(CombatError::BadRequest);
    }
    let condition = extract_string(body, "condition")
        .ok_or(CombatError::BadRequest)?
        .to_string();
    let duration = extract_int(body, "duration_rounds").ok_or(CombatError::BadRequest)?;
    if duration <= 0 {
        return Err(CombatError::BadRequest);
    }
    session
        .conditions
        .entry(target.clone())
        .or_default()
        .push(Condition {
            name: condition,
            remaining: duration,
        });
    let list = session.conditions.get(&target).unwrap();
    let conditions_json = list
        .iter()
        .map(|c| {
            format!(
                r#"{{"condition":"{}","remaining_rounds":{}}}"#,
                escape_json(&c.name),
                c.remaining
            )
        })
        .collect::<Vec<_>>()
        .join(",");
    Ok(format!(
        r#"{{"target":"{}","conditions":[{}]}}"#,
        escape_json(&target),
        conditions_json
    ))
}

/// Advance the turn tracker in a session.
///
/// Increments `turn_index` modulo the number of combatants; when the turn wraps
/// back to zero, `round` is incremented. Conditions on the newly active
/// combatant have their remaining duration decremented by one, and any
/// condition reaching zero is removed.
pub fn advance_turn(
    id: &str,
    sessions: &mut HashMap<String, CombatSession>,
) -> Result<String, CombatError> {
    let session = sessions.get_mut(id).ok_or(CombatError::NotFound)?;
    let len = session.order.len();
    if len == 0 {
        return Err(CombatError::BadRequest);
    }
    let new_index = (session.turn_index + 1) % len;
    if new_index == 0 {
        session.round += 1;
    }
    session.turn_index = new_index;
    let active_name = session.order[session.turn_index].name.clone();
    if let Some(list) = session.conditions.get_mut(&active_name) {
        list.retain(|c| c.remaining > 0);
        for c in list.iter_mut() {
            c.remaining -= 1;
        }
        list.retain(|c| c.remaining > 0);
    }
    let active = &session.order[session.turn_index];
    let conditions_json = format_conditions_map(session);
    Ok(format!(
        r#"{{"id":"{}","round":{},"turn_index":{},"active":{{"name":"{}","score":{}}},"conditions":{}}}"#,
        escape_json(&session.id),
        session.round,
        session.turn_index,
        escape_json(&active.name),
        active.score,
        conditions_json
    ))
}

/// Serialize all remaining conditions for a session as a JSON object keyed by
/// target name. Targets are sorted to keep the output deterministic.
fn format_conditions_map(session: &CombatSession) -> String {
    let mut targets: Vec<&String> = session.conditions.iter().map(|(k, _)| k).collect();
    targets.sort();
    let entries = targets
        .iter()
        .map(|t| {
            let list = session.conditions.get(*t).unwrap();
            let arr = list
                .iter()
                .map(|c| {
                    format!(
                        r#"{{"condition":"{}","remaining_rounds":{}}}"#,
                        escape_json(&c.name),
                        c.remaining
                    )
                })
                .collect::<Vec<_>>()
                .join(",");
            format!(r#""{}":[{}]"#, escape_json(t), arr)
        })
        .collect::<Vec<_>>()
        .join(",");
    format!("{{{}}}", entries)
}

/// Compute a single-use initiative order without creating a session.
pub fn handle_initiative(body: &str) -> Option<String> {
    let content = extract_array_content(body, "combatants")?;
    let order = roll_initiative(content)?;
    let order_json = order
        .into_iter()
        .map(|c| format!(r#"{{"name":"{}","score":{}}}"#, escape_json(&c.name), c.score))
        .collect::<Vec<_>>()
        .join(",");
    Some(format!(r#"{{"order":[{}]}}"#, order_json))
}
